src/Microsoft.OData.Client/DeepInsertSaveResult.cs

```diff
@@ -0,0 +1,242 @@
+        /// <param name="objects">Objects of the top-level type we want to perform a deep insert on.</param>
+        /// <exception cref="InvalidOperationException"></exception>
+        internal void BuildDescriptorGraph<T>(IEnumerable<Descriptor> descriptors, bool isRootObject, params T[] objects)
+        {
+            foreach (object obj in objects)
+            {
+                EntityDescriptor topLevelDescriptor = this.RequestInfo.EntityTracker.GetEntityDescriptor(obj);
+
+                // If we do not find an EntityDescriptor in the entitytracker
+                // for any of the provided objects then we throw an exception. 
+                if (topLevelDescriptor == null)
+                {
+                    throw Error.InvalidOperation(Strings.Context_EntityNotContained);
+                }
+
+                // If it's a top-level object sent by the user.
+                if (isRootObject)
+                {
+                    // Validate that we can only have one top level entity in Deep Insert.
+                    if (bulkUpdateGraph.TopLevelDescriptors.Count > 0)
+                    {
+                        throw Error.InvalidOperation(Strings.Context_DeepInsertOneTopLevelEntity);
+                    }
+
+                    bulkUpdateGraph.AddTopLevelDescriptor(topLevelDescriptor);
+                }
+
+                foreach (Descriptor descriptor in descriptors)
+                {
+                    if (descriptor is EntityDescriptor entityDescriptor)
+                    {
+                        if (entityDescriptor.ParentEntity != null &&
+                            entityDescriptor.ParentEntity.Equals(topLevelDescriptor.Entity) &&
+                            !bulkUpdateGraph.Contains(entityDescriptor))
+                        {
+                            // We don't support delete and update in deep insert.
+                            if (entityDescriptor.State == EntityStates.Deleted || entityDescriptor.State == EntityStates.Modified)
+                            {
+                                throw Error.InvalidOperation(Strings.Context_DeepInsertDeletedOrModified);
+                            }
+
+                            bulkUpdateGraph.AddRelatedDescriptor(topLevelDescriptor, entityDescriptor);
+                            this.BuildDescriptorGraph(descriptors, false, entityDescriptor.Entity);
+                        }
+                    }
+                    else if (descriptor is LinkDescriptor linkDescriptor)
+                    {
+                        if (linkDescriptor.Source.Equals(topLevelDescriptor.Entity) && !bulkUpdateGraph.Contains(linkDescriptor))
+                        {
+                            // We don't support delete and update in deep insert.
+                            if (linkDescriptor.State == EntityStates.Deleted || linkDescriptor.State == EntityStates.Modified)

```

Comment:
I think we need to add another check for SaveResultProcessed to check if the descriptor has been processed or not..

Reply:
In BaseSaveResult, we initialize `ChangedEntries`, which we use to create the request payload.

```csharp
this.ChangedEntries = context.EntityTracker.Entities.Cast<Descriptor>()
                                      .Union(context.EntityTracker.Links.Cast<Descriptor>())
                                      .Union(context.EntityTracker.Entities.SelectMany(e => e.StreamDescriptors).Cast<Descriptor>())
                                      .Where(o => o.IsModified && o.ChangeOrder != UInt32.MaxValue)
                                      .OrderBy(o => o.ChangeOrder)
                                      .ToList();
```

The `ChangedEntries` contain descriptors whose state `IsModified` == `true` so we ignore the `Unchanged` descriptors.
Once a result is processed, the descriptor state is changed to `Unchanged`. This ensures that the descriptor is not processed again in a subsequent `SaveChanges()` request.

In DeepInsert, we initialize the `BulkUpdateGraph` using the `ChangedEntries`.

```csharp
internal void DeepInsertRequest<T>(T resource)
{
    BuildDescriptorGraph(this.ChangedEntries, true, resource);

    .....

}
```

Example:
If we have delete and update operations which we then call SaveChanges(), those descriptors will not be part of the subsequent DeepInsert request.

Reply:
I will add E2E tests in the Deep insert response PR